check/cml-check-all.1
.\"
.\"  gcml2 -- an implementation of Eric Raymond's CML2 in C
.\"  Copyright (C) 2000-2002 Greg Banks
.\"
.\"  This library is free software; you can redistribute it and/or
.\"  modify it under the terms of the GNU Library General Public
.\"  License as published by the Free Software Foundation; either
.\"  version 2 of the License, or (at your option) any later version.
.\"
.\"  This library is distributed in the hope that it will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\"  Library General Public License for more details.
.\"
.\"  You should have received a copy of the GNU Library General Public
.\"  License along with this library; if not, write to the Free
.\"  Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.TH CML-CHECK-ALL "1" "August 2002" "GCML2" "GCML2 Manual Pages"
.\"
.\"
.SH NAME
cml\-check\-all \- wrapper for cml\-check and cml\-summarize
.\"
.\"
.SH SYNOPSIS
\fBcml\-check\-all\fR [\fIOPTION\fR]... \fIlinux-source-dir\fR
.\"
.\"
.SH DESCRIPTION
.PP
\fBcml\-check\-all\fR is a wrapper which can be used to run \fBcml\-check\fR
over multiple architecture config trees and summarize the result.  By default,
it runs \fBcml\-check\fR with the maximal useful set of warnings over all
the arch trees found in the given \fIlinux-source-dir\fR, then once in merge
mode, and provides a summary at verbosity level 0 (see \fBcml\-summarize\fR(1)
for a discussion of verbosity levels).
.PP
\fBcml\-check\-all\fR options are standard GNU long form and unless otherwise
stated those that take arguments may be used both with and without an = sign,
for example \fB\-\-no\-arch 386\fR or \fB\-\-no\-arch=386\fR.
.TP
\fB\-\-help\fR
display a summary of usage and exit.
.TP
\fB\-\-version\fR
display the version number and exit.
.TP
\fB\-\-no-arch\fR \fIarch\fR[,\fIarch\fR...]
Skip the given architecture; multiple architectures can be given,
separated by commas.  
.TP
\fB\-\-no-merge\fR
Skip the merge mode step.
.TP
\fB\-\-raw\fR
Skip the summary step and provide raw output from \fBcml\-check\fR.
.TP
\fB\-\-verbose\fR, \fB\-\-verbose=\fIlevel\fR
Passed to \fBcml\-summarize\fR.
.TP
\fB\-W\fIwarning\-name\fR, \fB\-Wno-\fIwarning\-name\fR, \fB\-Wall\fR, \fB\-Wno-all\fR
Passed to \fBcml\-check\fR.
.\"
.\"
.SH "REPORTING BUGS"
Report bugs to <gnb@alphalink.com.au>.
.\"
.\"
.SH "SEE ALSO"
.BR cml\-check (1),
.BR cml\-summarize (1).
.\"
.\"
.SH COPYRIGHT
Copyright \(co 2000-2002 Greg Banks.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


ChangeLog
Tue Sep 03 23:58:00 2002  Greg Banks <gnb@alphalink.com.au>
	* Release 0.7.1

	* Fixed memory corruption error where the banner node was being
	  freed but not removed from the hashtable of all nodes.
	* The CONFIG_DECSTATION bug in the mips port triggered an assert
	  failure in the overlap code; this case is now handled more
	  gracefully.
	* Added two new warnings, condition-loop and dependency-loop.
	* Gracefully handle case where a define_bool is conditional on itself.
	* Some recursive expression functions now check for recursion loops
	  and print an error message instead of blowing the stack.
	* Fixed misfeature where failure to find a "source"d file would
	  terminate parsing the remainder of the rulebase.
	* Using a symbol name instead of a sub-prompt as the default
	  value for a "choice" no longer causes a parse error.
	* Plugged a number of memory leaks in the CML1 parser.
	* Performance improvement for debug build; lots of spurious calls to
	  expr_as_string() avoided.
	* Added "save" debug keyword.
	* Warning message for mixed overlaps includes the overlap type.
	* Added some primitive memory-overrun debugging.
	* Removed some small cruft.
	* Library link order was broken.
	* Makefiles allow building against specific glib, not just the one
	  whose glib-config was found in $PATH.
	* Location of the shell for shell scripts is discovered at build time.
	* Added debugging and profiling support to cml-check-all shellscript.
	* A cml-check profile build will record elapsed time for each run and
	  save them to a file.
	* Fixed some warnings found by non-gcc compilers.
	* Replaced the variadic DDPRINTF() macro with a family of fixed
	  macros to enable compilation on c89 compilers.

Sun Aug 18 13:44:00 2002  Greg Banks <gnb@alphalink.com.au>
    	* Release 0.7.

	* Warnings can be individually enabled.  Default set depends on whether 
	  the parser is in merge mode.  API functions to enable/disable
	  warnings by index and convert a string name to an index.
	* Added check for define to nonliteral expression.
	* More precise check for ambiguous comparison against "n" only complains
	  for symbols which are forward-declared at the point when compared.
	* Made inconsistent-tag warnings more precise; doesn't emit spurious
	  warnings about define_bool not having an (EXPERIMENTAL) tag.
	* More precise check for forward references and forward dependencies can
	  tell the difference between forward and undeclared, at the cost of
	  some storage.
	* Check for overlapping definitions by reducing conditions to
	  disjunctive normal forms.
	* Added check for forward dependencies.
	* Added check for misuse of and dependency on arch-constant symbols
	  like CONFIG_X86.
	* Renamed summarise-warnings.awk -> summarize.awk and installed it.
	* Added cml-summarize shell script, which runs summarize.awk.
	* Added cml-check-all shell script, based on old dochecks.sh, but
	  now also handles running cml-summarize.
	* Added manpages for cml-check, cml-check-all and cml-summarize.
	  Description of errors and warnings for the cml-check manpage is
	  controlled in HTML and shared with the web page.
	* RPM package support: added spec file, rpm: target.

	* Added --version option to cml-check.
	* Fixed some compiler warnings that only showed up when compiling
	  with Mandrake's RPM_OPT_FLAGS.
	* cml-check enters merge mode only if more than 1 filename given.
	* Merged in TODO items from a separate TODO file that has been hanging
	  around for some time.
	* Centralised code for merging exprs to dependency rule exprs,
	  visibility exprs, and saveability exprs.
	* Allow enough room for state strings in node_gui_default_format_label.
	* Removed the toggle controlling whether node warnings are shown, as
	  that functionality has been merged into the state strings.
	* Added WEAK_POSITION flag to allow a node to be moved from its first
	  appearance in the tree as a define_bool to the expected place where it
	  appears as a bool using mn_reparent().
	* Separated out the calculation of branch treetype because it needs to
	  be used again when checking for mixed overlaps.
	* Reworked the code to merge the default exprs from various branches,
	  to deal with cases brought out by mixed-overlap testing.
	* TODO file lists specific tasks for bringing up to date with CML2
	  1.9.5 language spec (as if this will ever happen).
	* Cleaned up terminology: variable -> symbol.
	* Added mn_reparent() for moving nodes around in the menu tree.
	* There is now exactly one function to convert a branch type to a
	  string.


Tue Jul 23 00:15:00 2002  Greg Banks <gnb@alphalink.com.au>
	* Release 0.6
	* Main new feature is a new CML1 parser backend and a frontend
	  which reads all 17 CML1 arch trees into a unified tree and
	  does consistency checks.

	* Added CML1 parser.  All filenames related to parsing and lexing are
	  forked with CML1 and CML2 versions.
	* Renamed cm2lib/ directory to libcml/.
	* Renamed cml2lib.h to libcml.h.
	* Renamed libcml2.a to libcml.a
	* Renamed frontends from cml2-foo to cml-foo.
	* Automatically detects whether a rulebase filename is CML1 or CML2
	  and calls the appropriate parse routine.
	* Implemented CML1 parser error recovery, and implemented explicit error
	  messages for some errors, like unbalanced "endmenu".
	* Added merge mode: multiple arch trees are read into the same rulebase
	  and merged into a single tree, to find undefined symbols.
	* Added frontend to use merge mode for global syntax check.
	* Split parsing functionality of cml_rulebase_new() off into a new
	  function cml_rulebase_parse().  This allows some actions to be
	  taken between the creation and parsing, for example setting $ARCH.
	* CML1 parser handles "choice" statement where the default does
	  not match any of the sub-prompts.
	* Added a new expression operator, E_MDEP, which supports the strange
	  dependency semantics of the CML1 dep_mbool statement.

check/cml-check-all.sh
#!@BASH@

bindir=$(dirname $0)
[ "$bindir" = "." ] && bindir=$PWD

LINUXDIR=
CHECK="$bindir/cml-check"
CHECK_FLAGS=
SUMMARIZE="$bindir/cml-summarize"
SUMMARIZE_FLAGS=
DO_MERGE=yes
DO_SUMMARY=yes
NO_ARCHES="dummy,merge"
WARNINGS=

usage ()
{
    echo "Usage: $0 [--no-merge] [--no-arch ARCH[,ARCH...]] [--raw] linux-source-dir"
    exit ${1:-1}
}

fatal ()
{
    echo "$0: $*"
    exit 1
}

while [ $# -gt 0 ]; do
    case "$1" in
    --help) usage 0 ;;
    --version) echo "cml-check-all @VERSION@" ; exit 0 ;;
    --verbose) SUMMARIZE_FLAGS="$SUMMARIZE_FLAGS $1" ;;
    --verbose=*) SUMMARIZE_FLAGS="$SUMMARIZE_FLAGS $1" ;;
    --debug) CHECK_FLAGS="$CHECK_FLAGS $1 $2" ; shift ;;
    --debug=*) CHECK_FLAGS="$CHECK_FLAGS $1" ;;
    --no-merge) DO_MERGE= ;;
    --raw) DO_SUMMARY= ;;
    --no-arch) NO_ARCHES="$NO_ARCHES,$2" ; shift ;;
    --no-arch=*) NO_ARCHES="$NO_ARCHES,${1#*=}" ;;
    --warning) WARNINGS="$WARNINGS -W$2" shift ; ;;
    --warning=*) WARNINGS="$WARNINGS -W${1#*=}" ;;
    -W*) WARNINGS="$WARNINGS $1" ;;
    -*) usage 1 ;;
    *)
    	[ -z "$LINUXDIR" ] || usage 1
	LINUXDIR="$1"
	;;
    esac
    shift
done
[ -z "$LINUXDIR" ] && usage 1

# echo "LINUXDIR=$LINUXDIR"
# echo "CHECK=$CHECK"
# echo "CHECK_FLAGS=$CHECK_FLAGS"
# echo "SUMMARIZE=$SUMMARIZE"
# echo "SUMMARIZE_FLAGS=$SUMMARIZE_FLAGS"
# echo "DO_MERGE=$DO_MERGE"
# echo "NO_ARCHES=$NO_ARCHES"
# echo "WARNINGS=$WARNINGS"
# exit

[ -d "$LINUXDIR/arch" ] || fatal "${LINUXDIR}: No such directory"
ARCHES=$(cd $LINUXDIR/arch ; ls | egrep -v ^\("${NO_ARCHES//,/|}"\)\$)

if [ "$WARNINGS" ]; then
    CHECK_SINGLE_FLAGS="$WARNINGS"
    CHECK_MERGE_FLAGS="$WARNINGS"
else
    MERGE_WARNINGS="\
undeclared-symbol \
different-banner \
different-parent \
"

    CHECK_SINGLE_FLAGS="-Wall $(echo $MERGE_WARNINGS | sed -e 's|\([^[:blank:]]\+\)|-Wno-\1|g')"
    CHECK_MERGE_FLAGS="-Wno-all $(echo $MERGE_WARNINGS | sed -e 's|\([^[:blank:]]\+\)|-W\1|g')"
fi

# echo "CHECK_SINGLE_FLAGS=$CHECK_SINGLE_FLAGS"
# echo "CHECK_MERGE_FLAGS=$CHECK_MERGE_FLAGS"
# echo "ARCHES=$ARCHES"
# exit

dochecks ()
{
    [ "@PROFILE@" = "1" ] && rm cml-check.time
    for arch in $ARCHES ${DO_MERGE:+merge} ; do
	(
    	    cd $LINUXDIR
	    echo
	    echo "===== $arch"
	    if [ $arch = merge ]; then
		mode_flags="$CHECK_MERGE_FLAGS"
		target_arches="$ARCHES"
	    else
		mode_flags="$CHECK_SINGLE_FLAGS"
		target_arches=$arch
	    fi
    	    [ "@PROFILE@" = "1" -a -f gmon.out ] && rm -f gmon.out
    	    [ "@DEBUG@" = "1" -a -f core ] && rm -f core
	    $CHECK $CHECK_FLAGS \
	    	    $mode_flags \
	    	    $(echo $target_arches | sed -e 's|\([^[:blank:]]\+\)|arch/\1/config.in|g')
    	    [ "@PROFILE@" = "1" -a -f gmon.out ] && mv gmon.out gmon.$arch.out
    	    [ "@DEBUG@" = "1" -a -f core ] && mv core core.$arch
	)
    done

    # Merge the profile data into one
    (
	cd $LINUXDIR
	GMON_OUT=
	for arch in $ARCHES ${DO_MERGE:+merge} ; do
	    [ -f gmon.$arch.out ] && GMON_OUT="$GMON_OUT gmon.$arch.out"
	done
	gprof -s $CHECK $GMON_OUT
    	rm $GMON_OUT
    )

}

if [ "$DO_SUMMARY" ]; then
    dochecks 2>&1 | $SUMMARIZE $SUMMARIZE_FLAGS
else
    dochecks
fi

